controller/codebase.go

```diff
@@ -0,0 +1,453 @@
+	}
+	che := NewCheStarterClient(getClusterURL(), getNamespace(ctx))
+	workspace := WorkspaceRequest{
+		ID: ctx.WorkspaceID,
+	}
+	workspaceResp, err := che.CreateWorkspace(ctx, workspace)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"codebaseID": cb.ID,
+			"err":        err,
+		}, "unable to open workspaces")
+		if werr, ok := err.(*workspaceError); ok {
+			fmt.Println(werr.String())
+		}
+		return jsonapi.JSONErrorResponse(ctx, goa.ErrInternal(err.Error()))
+	}
+
+	resp := &app.WorkspaceOpen{
+		Links: &app.WorkspaceOpenLinks{
+			Open: &workspaceResp.WorkspaceIDEURL,
+		},
+	}
+	return ctx.OK(resp)
+}
+
+// CodebaseConvertFunc is a open ended function to add additional links/data/relations to a Codebase during
+// convertion from internal to API
+type CodebaseConvertFunc func(*goa.RequestData, *codebase.Codebase, *app.Codebase)
+
+// ConvertCodebases converts between internal and external REST representation
+func ConvertCodebases(request *goa.RequestData, codebases []*codebase.Codebase, additional ...CodebaseConvertFunc) []*app.Codebase {
+	var is = []*app.Codebase{}
+	for _, i := range codebases {
+		is = append(is, ConvertCodebase(request, i, additional...))
+	}
+	return is
+}
+
+// ConvertCodebase converts between internal and external REST representation
+func ConvertCodebase(request *goa.RequestData, codebase *codebase.Codebase, additional ...CodebaseConvertFunc) *app.Codebase {
+	codebaseType := APIStringTypeCodebase
+	spaceType := APIStringTypeSpace
+
+	spaceID := codebase.SpaceID.String()
+
+	selfURL := rest.AbsoluteURL(request, app.CodebaseHref(codebase.ID))
+	editURL := rest.AbsoluteURL(request, app.CodebaseHref(codebase.ID)+"/edit")
+	spaceSelfURL := rest.AbsoluteURL(request, app.SpaceHref(spaceID))
+
+	i := &app.Codebase{
+		Type: codebaseType,
+		ID:   &codebase.ID,
+		Attributes: &app.CodebaseAttributes{
+			CreatedAt: &codebase.CreatedAt,
+			Type:      &codebase.Type,
+			URL:       &codebase.URL,
+		},
+		Relationships: &app.CodebaseRelations{
+			Space: &app.RelationGeneric{
+				Data: &app.GenericData{
+					Type: &spaceType,
+					ID:   &spaceID,
+				},
+				Links: &app.GenericLinks{
+					Self: &spaceSelfURL,
+				},
+			},
+		},
+		Links: &app.CodebaseLinks{
+			Self: &selfURL,
+			Edit: &editURL,
+		},
+	}
+	for _, add := range additional {
+		add(request, codebase, i)
+	}
+	return i
+}
+
+// TODO: We need to dynamically get the real che namespace name from the tenant namespace from
+// somewhere more sensible then the token/generate/guess route.
+func getNamespace(ctx context.Context) string {
+	token := goajwt.ContextJWT(ctx)
+	if claims, ok := token.Claims.(jwt.MapClaims); ok {
+		email := claims["email"].(string)
+		return strings.Split(email, "@")[0] + "-dsaas-che"
+	}
+	return ""
+}
+
+func getClusterURL() *url.URL {
+	clusterURL, _ := url.Parse("https://tsrv.devshift.net:8443")
+	return clusterURL
+}
+
+// NewCheStarterClient is a helper function to create a new CheStarter client
```

Comment:
We do plan/code/build/test/deploy. Everything related to Che is "code". So, maybe "code" package? But it can be a bit confusing... "Code" can mean so many things :)